src/main/scala/is/hail/expr/ir/Compile.scala

```diff
@@ -0,0 +1,346 @@
+  }
+
+  private def present(x: Code[_]) = (const(false), x)
+  private def coerce[T](x: Code[_]): Code[T] = x.asInstanceOf[Code[T]]
+  private def tcoerce[T <: expr.Type](x: expr.Type): T = x.asInstanceOf[T]
+
+  def expression(ir: IR, fb: FunctionBuilder[_], env: E, mb: StagedBitSet): (Code[Boolean], Code[_]) = {
+    val region = fb.getArg[MemoryBuffer](1).load()
+    def expression(ir: IR, fb: FunctionBuilder[_] = fb, env: E = env, mb: StagedBitSet = mb): (Code[Boolean], Code[_]) =
+      Compile.expression(ir, fb, env, mb)
+    ir match {
+      case I32(x) =>
+        present(const(x))
+      case I64(x) =>
+        present(const(x))
+      case F32(x) =>
+        present(const(x))
+      case F64(x) =>
+        present(const(x))
+      case True() =>
+        present(const(true))
+      case False() =>
+        present(const(false))
+
+      case NA(typ) =>
+        (const(true), dummyValue(typ))
+      case IsNA(v) =>
+        present(expression(v)._1)
+      case MapNA(name, value, body, typ) =>
+        val vti = typeToTypeInfo(value.typ)
+        val bti = typeToTypeInfo(typ)
+        val (mvalue, vvalue) = expression(value)
+        val mx = mb.newBit()
+        fb.emit(mx := mvalue)
+        val x = fb.newLocal(name)(vti).asInstanceOf[LocalRef[Any]]
+        fb.emit(x := mx.mux(dummyValue(value.typ), vvalue))
+        val bodyenv = env.bind(name -> (vti, mx, x))
+        val (mbody, vbody) = expression(body, env = bodyenv)
+        (mvalue || mbody, mvalue.mux(dummyValue(typ), vbody))
+
+      case expr.ir.If(cond, cnsq, altr, typ) =>
+        val (mcond, vcond) = expression(cond)
+        val xvcond = mb.newBit()
+        fb.emit(xvcond := coerce[Boolean](vcond))
+        val (mcnsq, vcnsq) = expression(cnsq)
+        val (maltr, valtr) = expression(altr)
+
+        (mcond || (xvcond && mcnsq) || (!xvcond && maltr),
+          xvcond.asInstanceOf[Code[Boolean]].mux(vcnsq, valtr))
+
+      case expr.ir.Let(name, value, body, typ) =>
+        val vti = typeToTypeInfo(value.typ)
+        fb.newLocal(name)(vti) match { case x: LocalRef[v] =>
+          val (mvalue, vvalue) = expression(value)
+          val xmvalue = mb.newBit()
+          fb.emit(xmvalue := mvalue)
+          fb.emit(x := coerce[v](vvalue))
+          val bodyenv = env.bind(name -> (vti, xmvalue, x))
+          expression(body, env = bodyenv)
+        }
+      case Ref(name, typ) =>
+        val ti = typeToTypeInfo(typ)
+        val (t, m, v) = env.lookup(name)
+        assert(t == ti, s"$name type annotation, $typ, doesn't match typeinfo: $ti")
+        (m, v)
+
+      case ApplyPrimitive(op, args, typ) =>
+        val typs = args.map(_.typ)
+        val (margs, vargs) = args.map(expression(_)).unzip
+        val m = if (margs.isEmpty) const(false) else margs.reduce(_ || _)
+        (m, Primitives.lookup(op, typs, vargs))
+      case LazyApplyPrimitive(op, args, typ) =>
+        ???
+      case expr.ir.Lambda(names, body, typ) =>
+        ???
```

Comment:
Remove until we have uses/implementation?

Reply:
Remove the IRs or remove the branch from the match?

I'm not inclined to remove the branch because then Scala's missingness cannot help me.

Reply:
Both.  The IR compiler could go in soon and I see immediate uses for it even in its current state.  By taking these out, you leave no confusion to the user of the IR compiler what features are currently available and usable.

In other words, all things being equal, it is better if partial/staged changes that go into master are complete in-and-of-themselves (that is, self-contained partial changes) rather than leaving dangling bits that will get cleaned up/finished when later PRs finish the larger task of which the partial change is a part.

Reply:
Both gone.